### PR TITLE
Matrix stages to dynamically build Databricks shims [databricks]

### PR DESCRIPTION
**Pull request by NvTimLiu**
To fix [issue/8640
](https://github.com/NVIDIA/spark-rapids/issues/8640)
Matrix stages to dynamically build Databricks shims
    
To make the Databricks pre-merge CI easier to maintain, move Databricks CI script into a new file: Jenkinsfile.premerge-databricks
    
Then we only need to update the Databricks runtime index([10.4, 11.3, 12.2](https://github.com/NVIDIA/spark-rapids/compare/branch-23.08...NvTimLiu:spark-rapids:db-ci-matrix?expand=1#diff-6c8e5ccec30b82138e0f9b11481874f247e1a4b27e74b1e7f6261263eac89ce1R91)) in Jenkinsfile.premerge-databricks to add/remove the Databricks shims
    
Signed-off-by: Tim Liu <timl@nvidia.com>

### CODE REVIEW

**Comment by pxLi** [comment]
pre-merge now use PR merged commit jenkinsfile, so you should not need to make temp one to verify it.

feel free to clean the test code and run the CI, we can check logs later thanks~

File: jenkins/Jenkinsfile-blossom.premerge [hunk 1]
```diff
@@ -268,101 +257,27 @@ pipeline {
...
+                            try {
+                                build(job: 'tim-rapids-databricks_premerge-github1',
+                                    propagate: true, wait: true,
+                                    parameters: [
+                                            string(name: 'REF', value: params.REF),
+                                            string(name: 'GITHUB_DATA', value: params.GITHUB_DATA)
+                                    ])
+                            } catch (e){
+                                // 'JOB_NAME #ID completed with status FAILURE xxx' --> 'JOB_NAME/ID'
+                                def dbJobId = e.getCauses()[0].getShortDescription().split('completed with status FAILURE')[0].replace(' #', '/').trim()
```
Comment (pxLi) [inline jenkins/Jenkinsfile-blossom.premerge:272]: nit: can we try achieve this w/o exception catch ops? probably like

set `propagate: false` and use returned downstream_job (jenkins data model) to control the workflow
```
def DBJob = build(...)
if ( DBJob.result == 'SUCCESS' ) { // https://javadoc.jenkins.io/hudson/model/Result.html#field.summary
// ...
} else {
// do some ops use fields at https://javadoc.jenkins.io/hudson/model/Run.html?is-external=true#field.summary
}
```

**Comment by NvTimLiu** [comment]
build

**Comment by NvTimLiu** [comment]
build

File: jenkins/Jenkinsfile-blossom.premerge (hunk 1 above)
Comment (NvTimLiu) [reply jenkins/Jenkinsfile-blossom.premerge]: Good point, let me update it.

File: jenkins/Jenkinsfile-blossom.premerge-databricks [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        DATABRICKS_PUBKEY = credentials("SPARK_DATABRICKS_PUBKEY")
+        DATABRICKS_DRIVER = DbUtils.getDriver("$DB_TYPE")
+        DATABRICKS_WORKER = DbUtils.getWorker("$DB_TYPE")
+        INIT_SCRIPTS_DIR = "dbfs:/databricks/init_scripts/${BUILD_TAG}"
+    }
+
+    stages {
+        stage("Init githubHelper") {
+            steps {
+                script {
```
Comment (pxLi) [inline jenkins/Jenkinsfile-blossom.premerge-databricks:74]: please also help verify if we need to do some extra check if dup abort for the same PR.

seems the child job should be stopped if parent is abort, but need to double check thanks

**Comment by NvTimLiu** [comment]
build

File: jenkins/Jenkinsfile-blossom.premerge-databricks (hunk 2 above)
Comment (NvTimLiu) [reply jenkins/Jenkinsfile-blossom.premerge-databricks:74]: Yes, confirmed the child DB job also is aborted if parent is aborted

**Comment by NvTimLiu** [comment]
Failed DB log uploaed to the PR:[DB failed](https://github.com/NVIDIA/spark-rapids/actions/runs/5379765900/jobs/9761465291)

**Comment by NvTimLiu** [comment]
Create the issue https://github.com/NVIDIA/spark-rapids/issues/8640 to make this PR easy to understand and track

@jlowe  @tgravescs  @gerashegalov  @revans2  can you take a look when you have time, thanks!

**Review by revans2 — DISMISSED** [review]
Looks good to me. The config/code before lines up with the new config/code. The only problem is that I am not an expert on any of this so I would feel better if others also took a look.

**Review by jlowe — DISMISSED** [review]

File: jenkins/Jenkinsfile-blossom.premerge [hunk 3]
```diff
@@ -268,101 +257,24 @@ pipeline {
...
+                            if ( DBJob.result != 'SUCCESS' ) {
+                                // Output Databricks failure logs to uploaded onto the pre-merge PR
+                                print(DBJob.getRawBuild().getLog())
+                                // Fail the pipeline
+                                error "Databricks build result : " + DBJob.result
                             }
                         }
                     }
-                } // end of DB runtime 12.2
+                }// end of Databricks
```
Comment (gerashegalov) [inline jenkins/Jenkinsfile-blossom.premerge:277]: nit: comment
```suggestion
                } // end of Databricks
```

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by pxLi — DISMISSED** [review]

File: jenkins/Jenkinsfile-blossom.premerge (hunk 3 above)
Comment (NvTimLiu) [reply jenkins/Jenkinsfile-blossom.premerge]: Fixed

**Comment by NvTimLiu** [comment]
skip CI, the last commit only add a space before a code comment: https://github.com/NVIDIA/spark-rapids/pull/8608/commits/8e1f7b07edae22c318d82113f2441a20350a1ffd

**Comment by NvTimLiu** [comment]
build

**Review by GaryShen2008 — APPROVED** [review]

**Comment by NvTimLiu** [comment]
build

**Pull request merged**